nixos/qbittorrent-nox: init

Redhawk18 opened the pull request:
use tmpfiles instead of prescript

## Description of changes
This adds a service for the qbittorrent nox web ui with some basic options for the systemd job.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Redhawk18 committed:
nixos/qbittorrent-nox: init

use tmpfiles instead of prescript

suggestions

fmt

remove mddoc

refmt doc strings

add ability to declearive declear config

add description for settings option

fmt

fmt

add doc and remove settings

rename file

attempt to rebuild docs correctly

fmt

suggestions 3

doc suggestions

doc suggestion 2

doc suggestion 3

typo

doc suggestion 4

suggestions 4

fmt

suggestions 5

wants

suggestion 6

fmt

rename port and add hardening

fmt

misuzu commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,108 @@
+      '';
+    };
+
+    port = lib.mkOption {
```
Looks like this option does nothing

Redhawk18 replied:
Oops

misuzu commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,108 @@
+
+          ExecStart = "${cfg.package}/bin/qbittorrent-nox";
+          # To prevent "Quit & shutdown daemon" from working; we want systemd to manage it!
+          Restart = "on-success";
```
What if qbittorrent crashed?

Redhawk18 replied:
It was to stop, the quit option that would kill the service.

misuzu commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,108 @@
+      qbittorrent-nox = { gid = GID; };
+    };
+  };
+}
```
A doc would be great
```suggestion
  meta.doc = ./qbittorrent-nox.md;
}
```

Redhawk18 replied:
Where does this doc file go? Can I just inline it?

misuzu replied:
> Where does this doc file go? Can I just inline it?

`./qbittorrent-nox.md` here means `nixos/modules/services/torrent/qbittorrent-nox.md`

Redhawk18 replied:
waiting on final approval to write docs now.

misuzu reviewed (commented):
If I remember correctly, qbittorrent-nox will want you to accept some kind of legal agreement or it won't start. How is this supposed to be handled?

In [my module](https://gitlab.com/misuzu/nixos-configuration/-/blob/main/common/modules/nixosModules/qbittorrent/default.nix#L59) I can just set `services.qbittorrent.settings.LegalNotice.Accepted = true;`, something like this would be great. There could also be an assert like [this](https://github.com/NixOS/nixpkgs/blob/c374d94f1536013ca8e92341b540eba4c22f9c62/nixos/modules/services/games/minecraft-server.nix#L272).

Redhawk18 commented:
Is the uid and gid okay? I wasn't sure if there's a way to generate it dynamically.

Redhawk18 commented:
> If I remember correctly, qbittorrent-nox will want you to accept some kind of legal agreement or it won't start. How is this supposed to be handled?
> 
> In [my module](https://gitlab.com/misuzu/nixos-configuration/-/blob/50c9fc24da4b95c8f0349739688107dbba31a58e/common/modules/nixosModules/qbittorrent/default.nix#L18) I can just set `services.qbittorrent.settings.LegalNotice.Accepted = true;`, something like this would be great. There could also be an assert like [this](https://github.com/NixOS/nixpkgs/blob/c374d94f1536013ca8e92341b540eba4c22f9c62/nixos/modules/services/games/minecraft-server.nix#L272).

I believe the desktop app does, But I installed this to test it on my desktop and I was never prompted once to agree to it. It's likely a bug on their end, and even if we added an assert the eula would still show up. It's not like with minecraft where you could write `true` to a file and have it be accepted I believe.

Redhawk18 commented:
@NixOS/nix-formatting The error log told me to ping you with errors.

misuzu commented:
> Is the uid and gid okay? I wasn't sure if there's a way to generate it dynamically.

See https://github.com/NixOS/rfcs/blob/62d1245ec1eae275edb996229585637035c02538/rfcs/0052-dynamic-ids.md#using-ids-dynamically-allocated-by-nixos



> But I installed this to test it on my desktop and I was never prompted once to agree to it.

You need to run your module on a fresh machine. A NixOS test would actually be great way to make sure it works.



> even if we added an assert the eula would still show up. It's not like with minecraft where you could write `true` to a file and have it be accepted I believe.

Why not? Adding this `qBittorrent.conf` works just fine:
```
[LegalNotice]
Accepted=true
```
It's exactly what the module I linked does.


> @NixOS/nix-formatting The error log told me to ping you with errors.

The error log says this:

> Please go to the Nixpkgs root directory, run `nix-shell`, then: nixfmt 'nixos/modules/services/torrent/qbittorrent-nox.nix'

Redhawk18 commented:
I read the error message lol, I asked because I did what it suggested and it still showed an error, but it looks like it can't find the file to begin with.

Redhawk18 commented:
> > Is the uid and gid okay? I wasn't sure if there's a way to generate it dynamically.
> 
> See https://github.com/NixOS/rfcs/blob/62d1245ec1eae275edb996229585637035c02538/rfcs/0052-dynamic-ids.md#using-ids-dynamically-allocated-by-nixos
> 
> > But I installed this to test it on my desktop and I was never prompted once to agree to it.
> 
> You need to run your module on a fresh machine. A NixOS test would actually be great way to make sure it works.
> 
> > even if we added an assert the eula would still show up. It's not like with minecraft where you could write `true` to a file and have it be accepted I believe.
> 
> Why not? Adding this to the `qBittorrent.conf` works just fine:
> 
> ```
> [LegalNotice]
> Accepted=true
> ```
> 
> It's exactly what the module I linked does.
> 
> > @NixOS/nix-formatting The error log told me to ping you with errors.
> 
> The error log says this:
> 
> > Please go to the Nixpkgs root directory, run `nix-shell`, then: nixfmt 'nixos/modules/services/torrent/qbittorrent-nox.nix'

Will this override the config even if the settings option is not set?

misuzu commented:
> I read the error message lol, I asked because I did what it suggested and it still showed an error, but it looks like it can't find the file to begin with.

Not sure what you mean, it works fine:
```
% nix run nixpkgs#gh -- pr checkout -f 337109
% nix-shell
$ nixfmt 'nixos/modules/services/torrent/qbittorrent-nox.nix'
$ git diff
diff --git a/nixos/modules/services/torrent/qbittorrent-nox.nix b/nixos/modules/services/torrent/qbittorrent-nox.nix
index 55bd6805d4b7..cf2cfe5bdbbd 100644
--- a/nixos/modules/services/torrent/qbittorrent-nox.nix
+++ b/nixos/modules/services/torrent/qbittorrent-nox.nix
@@ -1,6 +1,13 @@
-{ config, lib, pkgs, ... }:
-let cfg = config.services.qbittorrent-nox;
-in {
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.qbittorrent-nox;
+in
+{
   # meta.doc = "./qbittorrent-nox.md"; # TODO remind me, since I rather write it later, than write now and rewrite later.

   options.services.qbittorrent-nox = {
@@ -9,15 +16,13 @@ in {
     dataDir = lib.mkOption {
       type = lib.types.path;
:...skipping...
```

Redhawk18 commented:
> nixfmt 'nixos/modules/services/torrent/qbittorrent-nox.nix'

Ah I ran `nix-shell -p nixfmt` I didn't know there was a specific version of nixfmt just for nixpkgs.

Redhawk18 commented:
Are the `meta.docs` what are shown on the nixos wiki?

infinisil commented:
Improving the message to hopefully prevent such confusion with https://github.com/NixOS/nixpkgs/pull/337577 :smile:

Redhawk18 commented:
> > Is the uid and gid okay? I wasn't sure if there's a way to generate it dynamically.
> 
> See https://github.com/NixOS/rfcs/blob/62d1245ec1eae275edb996229585637035c02538/rfcs/0052-dynamic-ids.md#using-ids-dynamically-allocated-by-nixos
> 
> > But I installed this to test it on my desktop and I was never prompted once to agree to it.
> 
> You need to run your module on a fresh machine. A NixOS test would actually be great way to make sure it works.
> 
> > even if we added an assert the eula would still show up. It's not like with minecraft where you could write `true` to a file and have it be accepted I believe.
> 
> Why not? Adding this to the `qBittorrent.conf` works just fine:
> 
> ```
> [LegalNotice]
> Accepted=true
> ```
> 
> It's exactly what the module I linked does.
> 
> > @NixOS/nix-formatting The error log told me to ping you with errors.
> 
> The error log says this:
> 
> > Please go to the Nixpkgs root directory, run `nix-shell`, then: nixfmt 'nixos/modules/services/torrent/qbittorrent-nox.nix'

Hey I did some testing on this idea and it's insanely buggy and hard to get working correctly. I'm likely going to remove the ability to add settings. There's also no docs explaining the options, so it leads me to believe that the qbittorrent team views this as an internal config file. Which means they will change how data is structured without warning which would mean all settings set by us would no longer work and there's no declarative way to really stop this.

misuzu commented:
> Are the `meta.docs` what are shown on the nixos wiki?

They are shown here https://nixos.org/manual/nixos/unstable/

misuzu reviewed (commented):
Added some suggestions.
Also found this pull request: https://github.com/NixOS/nixpkgs/pull/287923

Redhawk18 commented:
I think this pr is different enough, that other pr seems very focused on all the details of a declarative system and is inactive, this is just the ground work for this service. We can always submit a new pr with those options later.

misuzu reviewed (commented):
It works! Added some suggestions for the documentation

Redhawk18 commented on nixos/modules/services/torrent/qbittorrent-nox.md:
```diff
@@ -0,0 +1,20 @@
+
+## Configuration {#module-services-torrent-qbittorrent-nox-configuration}
+
+Options can be decleared like the following. Check the documentation for more options.
```
I was referring to this website
https://search.nixos.org/options

misuzu replied:
You can add a link to https://search.nixos.org/options?query=services.qbittorrent

Redhawk18 commented on nixos/modules/services/torrent/qbittorrent-nox.md:
````diff
@@ -0,0 +1,20 @@
+
+Options can be decleared like the following. Check the documentation for more options.
+```nix
+services.qbittorrent-nox = {
+  enable = true;
+
+  # This directory is where application info is saved.
+  dataDir = "path/to/share"
+
+  user = "root";
+  openFirewall = true;
+};
````
Is it possible to add these docs to the nixos wiki? I have personally never used the manual and didn't even know these docs would be included in that.

misuzu replied:
Without this basic information your module is not beginner friendly. Number one complaint about NixOS is the lack of documentation, and I'd like this to change.
You can add a wiki page later, if you want add something more abount your module, but I don't think that would be necessary.

misuzu commented:
Right, you should also add a release notes entry for your module.

misuzu commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -91,6 +91,8 @@
 
 - [Proton Mail bridge](https://proton.me/mail/bridge), a desktop application that runs in the background, encrypting and decrypting messages as they enter and leave your computer. It lets you add your Proton Mail account to your favorite email client via IMAP/SMTP by creating a local email server on your computer.
 
+- [qBittorrent-nox](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer)), a headless web ui for qBittorrent. That allows a torrent client to run as a daemon on your computer.
```
```suggestion
- [qBittorrent-nox](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer)), a headless web ui for qBittorrent. That allows a torrent client to run as a daemon on your computer. Available as [services.qbittorrent-nix](#opt-services.qbittorrent-nix.enable).
```

misuzu commented on nixos/modules/services/torrent/qbittorrent-nox.md:
```diff
@@ -0,0 +1,17 @@
+
+qBittorrent-nox is available as a standalone package as `qbittorrent-nox`, it can be installed to test the program before setting up the service.
+
+## Configuration {#module-services-torrent-qbittorrent-nox-configuration}
+
```
```suggestion
```

This can be removed to avoid one-entry "Table of Contents"

misuzu commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -91,6 +91,8 @@
 
 - [Proton Mail bridge](https://proton.me/mail/bridge), a desktop application that runs in the background, encrypting and decrypting messages as they enter and leave your computer. It lets you add your Proton Mail account to your favorite email client via IMAP/SMTP by creating a local email server on your computer.
 
+- [qBittorrent-nox](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer)), a headless web ui for qBittorrent. That allows a torrent client to run as a daemon on your computer. Available as [services.qbittorrent-nix](#opt-services.qbittorrent-nix.enable).
```
There's a typo:
```suggestion
- [qBittorrent-nox](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer)), a headless web ui for qBittorrent. That allows a torrent client to run as a daemon on your computer. Available as [services.qbittorrent-nox](#opt-services.qbittorrent-nox.enable).
```

Redhawk18 replied:
oops

Redhawk18 commented:
I hope thats it!

misuzu commented on nixos/modules/services/torrent/qbittorrent-nox.md:
```diff
@@ -0,0 +1,15 @@
+
+  port = 8080;
+  openFirewall = true;
+  user = "root";
```
Since this will likely be copy-pasted, it's better to either remove this line or replace it with a specific username.

Redhawk18 replied:
I disagree, I added it here because I had problems with sonarr being unable to have permissions to hardlink files because it did not have rw permissions to qbittorrent's downloads folder. For sure I have a skill issue with linux permissions but having everything under root makes it easier at first.

misuzu replied:
Running things as root must be discouraged as much as possible, especially in the official documentation.

misuzu reviewed (approved):
##### Reviewed points

- [X] module path fits the guidelines
- [ ] module tests succeed on ARCHITECTURE (no NixOS test)
- [X] options have appropriate types
- [X] options have default
- [ ] options have example (no need here imo)
- [X] options have descriptions
- [X] no unneeded package is added to `environment.systemPackages`
- [X] `meta.maintainers` is set
- [X] module documentation is declared in `meta.doc`
- [X] a release notes entry for a module is present

##### Possible improvements

I'm not sure the `dataDir` option is actually useful. On my qBittorrent instance, which has around 500 active torrents, the profile directory takes only 42M of disk space. What actually would be useful is an option to change the default downloads directory.

Also a NixOS test would be great, but it's optional.

##### Comments

Looks good enough to me!

Redhawk18 commented:
> ##### Reviewed points
> 
>     * [x]  module path fits the guidelines
> 
>     * [ ]  module tests succeed on ARCHITECTURE (no NixOS test)
> 
>     * [x]  options have appropriate types
> 
>     * [x]  options have default
> 
>     * [ ]  options have example (no need here imo)
> 
>     * [x]  options have descriptions
> 
>     * [x]  no unneeded package is added to `environment.systemPackages`
> 
>     * [x]  `meta.maintainers` is set
> 
>     * [x]  module documentation is declared in `meta.doc`
> 
>     * [x]  a release notes entry for a module is present
> 
> 
> ##### Possible improvements
> 
> I'm not sure the `dataDir` option is actually useful. On my qBittorrent instance, which has around 500 active torrents, the profile directory takes only 42M of disk space. What actually would be useful is an option to change the default downloads directory.
> 
> Also a NixOS test would be great, but it's optional.
> ##### Comments
> 
> Looks good enough to me!

What kind of test would you want sense this a a ui? Also I agree that datadir is a bad option, but currently I'm not a nix god with merging options into config files and theres little to no docs for those options and how qbittorrent internally organizes/names them. So I wanted to skip that rabbit hole for this pr at least. It seems #287923 is working on that so maybe they can rebase ontop of this pr.

misuzu commented:
> What kind of test would you want sense this a a ui?

Something like this: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/bittorrent.nix

Redhawk18 commented:
> > What kind of test would you want sense this a a ui?
> 
> Something like this: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/bittorrent.nix

I don't think I'm advanced enough to write a test like that yet.

misuzu commented:
> I don't think I'm advanced enough to write a test like that yet.

Like I said, the test is optional 😁

Redhawk18 commented:
> > I don't think I'm advanced enough to write a test like that yet.
> 
> Like I said, the test is optional 😁

Thanks, I believe this is ready for review now.

nevivurn commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,102 @@
+
+          ExecStart = "${cfg.package}/bin/qbittorrent-nox --webui-port=${toString cfg.port}";
+
+          IOSchedulingClass = "idle";
+          IOSchedulingPriority = "7";
+
```
Just a question, what effect does this have?

Redhawk18 replied:
Like all scheduling, either massive or no changes. But it seems like a nice to have.

Redhawk18 commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,102 @@
+      services.qbittorrent-nox = {
+        description = "qBittorrent-nox service";
+        documentation = [ "man:qbittorrent-nox(1)" ];
+        after = [ "network.target" ];
```
[We have more options and a better system than they do](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer))

nevivurn replied:
Yeah, I think not using the upstream unit is completely fine, I was referring to the unit dependencies only.
I believe these need fixing, `After=network.target` is insufficient in my experience.

```
Wants=network-online.target
After=network-online.target nss-lookup.target
```

is what upstream ships (actually, they also include `After=local-fs.target` but that's unnecessary afaik) & what they recommend in the documentation you linked, unless I'm misunderstanding you.

Redhawk18 replied:
Why is it unneeded?

nevivurn replied:
local-fs.target should be already pulled in by all services unless `DefaultDependencies=no`.
We are still missing `Wants=network-online.target`.

From `systemd.special(7)`:

> Units that strictly require a configured network connection should pull in network-online.target (via a Wants= type dependency) and order themselves after it. This target unit is intended to pull in a service that delays further execution until the network is sufficiently set up. What precisely this requires is left to the implementation of the network managing service.

bbigras commented:
I got this error:

```
       error: The option `systemd.services.qbittorrent-nox.Wants' does not exist. Definition values:
       - In `/nix/store/k1z1bn2qrkdj7yvd1s5b7r93h1ygnbxc-source/nixos/modules/services/torrent/qbittorrent-nox.nix':
           [
             "network-online.target"
           ]
```

Redhawk18 commented:
> I got this error:
> 
> ```
>        error: The option `systemd.services.qbittorrent-nox.Wants' does not exist. Definition values:
>        - In `/nix/store/k1z1bn2qrkdj7yvd1s5b7r93h1ygnbxc-source/nixos/modules/services/torrent/qbittorrent-nox.nix':
>            [
>              "network-online.target"
>            ]
> ```
> 
> is `Wants` supposed to be lowercase?

All systemd "headers" should be uppercase

> I got this error:
> 
> ```
>        error: The option `systemd.services.qbittorrent-nox.Wants' does not exist. Definition values:
>        - In `/nix/store/k1z1bn2qrkdj7yvd1s5b7r93h1ygnbxc-source/nixos/modules/services/torrent/qbittorrent-nox.nix':
>            [
>              "network-online.target"
>            ]
> ```
> 
> is `Wants` supposed to be lowercase?

Yeah systemd and nix mix upper and lower cases for extra fun

Redhawk18 commented:
Please rereview @bbigras

bbigras reviewed (approved):
It runs fine :)

Thank you very much for the module.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4514

SuperSandro2000 commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+          User = cfg.user;
+          Group = cfg.group;
+
+          ExecStart = "${cfg.package}/bin/qbittorrent-nox --webui-port=${toString cfg.port}";
```
We could use lib.getExe here

Redhawk18 commented:
@SuperSandro2000

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+    maintainers = with lib.maintainers; [ redhawk ];
+  };
+
+  options.services.qbittorrent-nox = {
```
IMO the service should be called `qbittorent`.

Redhawk18 replied:
No, qbittorrent is a completely different beast. I do not believe you can run it headless like you can with the webui.

ambroisie replied:
I'm not saying you can (though I think it'd be easier than you might think, on Xorg you probably just need to import the `DISPLAY` env variable)

My point is this service is to used to run/host qBittorent. That it is running the `-nox` package is an implementation detail, and should _not_ be surfaced in the service name.

Redhawk18 replied:
> I'm not saying you can (though I think it'd be easier than you might think, on Xorg you probably just need to import the `DISPLAY` env variable)
> 
> My point is this service is to used to run/host qBittorent. That it is running the `-nox` package is an implementation detail, and should _not_ be surfaced in the service name.

Until an actual user wants to do this lets leave it be for now, you're overthinking this. Also isn't xorg entering a deprecation phase? Why would we add more support for a project it's own developers don't even want to work on.

ambroisie replied:
Again, I'm not saying we need to figure out how to run the GUI version (that's not relevant). But the naming is an issue, I still think this should be renamed before getting merged.

Redhawk18 replied:
And I still disagree, just because the main qbittorrent package is named that and split into 3 packages does not mean this service should be named after it. If this service only works for qbittorrent nox it should be named after qbittorrent nox. It is HIGHLY relevant here, you are mistaken.

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+  };
+
+  options.services.qbittorrent-nox = {
+    enable = lib.mkEnableOption "qBittorrent-nox headless";
```
```suggestion
    enable = lib.mkEnableOption "qBittorrent service";
```

Redhawk18 replied:
Can you prove that qbittorrent with the qt desktop ui can run headless?

ambroisie replied:
That's not what I'm saying.

Redhawk18 replied:
> That's not what I'm saying.

See my response above.

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+
+    dataDir = lib.mkOption {
+      type = lib.types.path;
+      default = "/var/lib/qbittorrent-nox";
```
```suggestion
      default = "/var/lib/qbittorrent";
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+    dataDir = lib.mkOption {
+      type = lib.types.path;
+      default = "/var/lib/qbittorrent-nox";
+      description = "The directory where qBittorrent-nox stores its data files.";
```
```suggestion
      description = "The directory where qBittorrent stores its data files.";
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+    openFirewall = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = "Open services.qbittorrent-nox.port to the outside network.";
```
```suggestion
      description = "Open qBittorent ports to the outside network.";
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+
+    user = lib.mkOption {
+      type = lib.types.str;
+      default = "qbittorrent-nox";
+      description = "User account under which qBittorrent-nox runs.";
```
```suggestion
      default = "qbittorrent";
      description = "User account under which qBittorrent runs.";
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+
+    group = lib.mkOption {
+      type = lib.types.str;
+      default = "qbittorrent-nox";
+      description = "Group under which qBittorrent-nox runs.";
```
```suggestion
      default = "qbittorrent";
      description = "Group under which qBittorrent runs.";
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+    port = lib.mkOption {
+      type = lib.types.port;
+      default = 8080;
+      description = "qBittorrent-nox web UI port.";
```
```suggestion
      description = "qBittorrent web UI port.";
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+      description = "qBittorrent-nox web UI port.";
+    };
+
+    package = lib.mkPackageOption pkgs "qbittorrent-nox" { };
```
```suggestion
    package = lib.mkPackageOption pkgs "qBittorrent" { default = "qbittorrent-nox" };
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+    networking.firewall = lib.mkIf cfg.openFirewall { allowedTCPPorts = [ cfg.port ]; };
+
+    systemd = {
+      services.qbittorrent-nox = {
+        description = "qBittorrent-nox service";
+        documentation = [ "man:qbittorrent-nox(1)" ];
```
```suggestion
      services.qbittorrent = {
        description = "qBittorrent service";
        documentation = [ "man:qbittorrent-nox(1)" ];
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+        environment.QBT_PROFILE = cfg.dataDir; # required
+      };
+
+      tmpfiles.settings."qbittorrent-nox" = {
```
```suggestion
      tmpfiles.settings."qbittorrent" = {
```

ambroisie commented on nixos/modules/services/torrent/qbittorrent-nox.nix:
```diff
@@ -0,0 +1,106 @@
+      };
+    };
+
+    users = {
+      users = lib.mkIf (cfg.user == "qbittorrent-nox") {
+        qbittorrent-nox = {
+          group = cfg.group;
+          isSystemUser = true;
+        };
+      };
+
+      groups = lib.mkIf (cfg.group == "qbittorrent-nox") {
+        qbittorrent-nox = { };
+
+      };
+    };
```
```suggestion
    users = {
      users = lib.mkIf (cfg.user == "qbittorrent") {
        qbittorrent = {
          group = cfg.group;
          isSystemUser = true;
        };
      };

      groups = lib.mkIf (cfg.group == "qbittorrent") {
        qbittorrent = { };
      };
    };
```

ambroisie commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -91,6 +91,8 @@
 
 - [Proton Mail bridge](https://proton.me/mail/bridge), a desktop application that runs in the background, encrypting and decrypting messages as they enter and leave your computer. It lets you add your Proton Mail account to your favorite email client via IMAP/SMTP by creating a local email server on your computer.
 
+- [qBittorrent-nox](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer)), a headless web ui for qBittorrent. That allows a torrent client to run as a daemon on your computer. Available as [services.qbittorrent-nox](#opt-services.qbittorrent-nox.enable).
```
```suggestion
- [qBittorrent](https://github.com/qbittorrent/qBittorrent/wiki/Running-qBittorrent-without-X-server-(WebUI-only,-systemd-service-set-up,-Ubuntu-15.04-or-newer)), a headless web ui for qBittorrent. That allows a torrent client to run as a daemon on your computer. Available as [services.qbittorrent](#opt-services.qbittorrent.enable).
```

ambroisie reviewed (changes requested):
The `port` option is going to be ambiguous when someone adds an option for `--torrenting-port`.

In general, I think #287923 is in a slightly better shape (with declarative config mostly done, and hardening options on the service). I would rather we merge the parts that are already working from that PR rather than deviate more.

Redhawk18 commented:
> The `port` option is going to be ambiguous when someone adds an option for `--torrenting-port`.
> 
> In general, I think #287923 is in a slightly better shape (with declarative config mostly done, and hardening options on the service). I would rather we merge the parts that are already working from that PR rather than deviate more.

I highly disagree, and blocking my pr seems incredibility hostile to me. Options like a torrenting port will be hidden within a nested ini settings page there is no cli torrenting port option
```bash
Usage:
    qbittorrent-nox [options] [(<filename> | <url>)...]
Options:
    -v | --version             Display program version and exit
    -h | --help                Display this help message and exit
    --webui-port=<port>        Change the WebUI port
    --torrenting-port=<port>   Change the torrenting port
    -d | --daemon              Run in daemon-mode (background)
    --profile=<dir>            Store configuration files in <dir>
    --configuration=<name>     Store configuration files in directories
                               qBittorrent_<name>
    --relative-fastresume      Hack into libtorrent fastresume files and make
                               file paths relative to the profile directory
    files or URLs              Download the torrents passed by the user

Options when adding new torrents:
    --save-path=<path>         Torrent save path
    --add-paused=<true|false>  Add torrents as started or paused
    --skip-hash-check          Skip hash check
    --category=<name>          Assign torrents to category. If the category
                               doesn't exist, it will be created.
    --sequential               Download files in sequential order
    --first-and-last           Download first and last pieces first
    --skip-dialog=<true|false> Specify whether the "Add New Torrent" dialog
                               opens when adding a torrent.

Option values may be supplied via environment variables. For option named
'parameter-name', environment variable name is 'QBT_PARAMETER_NAME' (in upper
case, '-' replaced with '_'). To pass flag values, set the variable to '1' or
'TRUE'. For example, to disable the splash screen: 
QBT_NO_SPLASH=1 qbittorrent-nox
Command line parameters take precedence over environment variables
```
Like I already said in the other pr @fsnkty admitted they are unable to work on this and the pr is incomplete. I rather have a working service that doesn't have all the options it possibility could than a pr that has been open for years that is almost perfect. My pr is finished and already has 4 reviews, blocking it at this point seems like whoever was first gets to merge their pr. Not whoever has the better and more active pr with more approval.

ambroisie commented:
@Redhawk18 I'm not blocking your PR, I'm pointing out that both PRs are to add essentially the same module. I did see that @fsnkty is not currently able to work on their PR, and I do appreciate you wanting to get _something_ merged in.

But from comparing both PRs, I think yours is diverging from the first one for no reason that I can see, and I would rather we just have the same basis as I think the first PR had more reviews done to it and resolved some of those same pain points. 

I said:

> I would rather we merge the parts that are already working from that PR

This might have been phrased badly, but I hope my answer cleared up than I'm not blocking this PR. I _do_ want to see it be merged, but I voiced what I think is wrong in my review (namely, the naming of the module, the `port` option, and the hardening).

Redhawk18 commented:
> This might have been phrased badly, but I hope my answer cleared up than I'm not blocking this PR. I _do_ want to see it be merged, but I voiced what I think is wrong in my review (namely, the naming of the module, the `port` option, and the hardening).

Fine, I renamed the port and added the hardening, but I'm not renaming the module. If you can prove qbittorrent can run over xorg which is on life support first off and work on wayland I will rename the module. The name is correct how it is now because it only uses nox, so it should be named after it.

fsnkty commented:
>  blocking my pr seems incredibility hostile to me. 

nothings personal or being blocked.

> Like I already said in the other pr @fsnkty admitted they are unable to work on this and the pr is incomplete.

it needs testing and some minor fixup, it's been in use by multiple people for months now including myself, qbit is a bizarre program and I simply didn't want to have the module included just for it to break in bizarre ways due to some qbit feature I was unaware of.
it's simply nicer to get stuff right in the first PR instead of messing about with stages of them and messing about with the config/settings of what people already have done.
I hope you understand it well

> I rather have a working service that doesn't have all the options it possibility could than a pr that has been open for years that is almost perfect. 

take what works 100% then, no sense in nearly duplicating work and rerequesting reviews, it also creates more work for anyone who now has to work around and refine the module you've made, or argue for a change in it's config

> My pr is finished and already has 4 reviews, blocking it at this point seems like whoever was first gets to merge their pr. Not whoever has the better and more active pr with more approval.

not being blocked, just please be cautious of the extra work this causes for yourself and for anyone else who wants to work on the module as it'd now make extra changes
that PR of mine has some discussion on how to handle things like port config/setting and has afaict decent structure and good practices on stuff like service hardening, please take that into consideration at the very least, let's avoid creating extra work 😞

Redhawk18 commented:
> > blocking my pr seems incredibility hostile to me.
> 
> nothings personal or being blocked. 

I still believe blocking my pr over a minor name change that implies this service works with both the ui and headless version of qbittorrent is a bad idea.

> take what works 100% then, no sense in nearly duplicating work and rerequesting reviews, it also creates more work for anyone who now has to work around and refine the module you've made, or argue for a change in it's config

I had no idea until the pr had progressed far enough to be reviewed. It seems my pr has made your pr more active as well by the threat of being closer to merging. And I understand you don't want to edit my code, and rebase your pr which I understand. I'd also rather write everything myself aswell. I made this pr because I wanted this service and you were inactive, I understand you can't choose when you have time in your life for free unpaid volunteer work that gets no credit anywhere. But I believe I have been incredibility reasonable with all suggestions, and everything was progressing smoothly until now.

The reason a block offenses me so much is the amount of work I've put into this for someone on the internet to just click a reject button with a bad suggestion, and prevent this from being merged is insane to me.

fsnkty commented:
> > > blocking my pr seems incredibility hostile to me.
> > 
> > nothings personal or being blocked. 
> 
> I still believe blocking my pr over a minor name change that implies this service works with both the ui and headless version of qbittorrent is a bad idea.
> 
> > take what works 100% then, no sense in nearly duplicating work and rerequesting reviews, it also creates more work for anyone who now has to work around and refine the module you've made, or argue for a change in it's config
> 
> I had no idea until the pr had progressed far enough to be reviewed. It seems my pr has made your pr more active as well by the threat of being closer to merging. And I understand you don't want to edit my code, and rebase your pr which I understand. I'd also rather write everything myself aswell. I made this pr because I wanted this service and you were inactive, I understand you can't choose when you have time in your life for free unpaid volunteer work that gets no credit anywhere. But I believe I have been incredibility reasonable with all suggestions, and everything was progressing smoothly until now.
> 
> The reason a block offenses me so much is the amount of work I've put into this for someone on the internet to just click a reject button with a bad suggestion, and prevent this from being merged is insane to me.
> 
> 
> 

it's not about my time, the differences here seem unnecessary, it's not about wanting to avoid rebasing.
blindly calling the advice of others to be a bad suggestion isn't productive.
just looking to save us all some time n effort ^^```` but responses like this are more frustrating than productive, I'll leave you to it but please don't let this enable you to hold up the other PR later on whenever it's ready.. I'd still like to just use it in it's entirety rather than mushing this seemingly pointlessly diverged work together

Redhawk18 commented:
> it's not about my time, the differences here seem unnecessary, it's not about wanting to avoid rebasing. blindly calling the advice of others to be a bad suggestion isn't productive. just looking to save us all some time n effort ^^```` but responses like this are more frustrating than productive, I'll leave you to it but please don't let this enable you to hold up the other PR later on whenever it's ready.. I'd still like to just use it in it's entirety rather than mushing this seemingly pointlessly diverged work together

What do you want to call a suggestion I strongly disagree with, that blocks my entire pr and makes my work a waste of time? I came up with this pr by myself as something I wanted to improve to the nix ecosystem. There's no diverging of work, just because you made the pr first doesn't mean my work is a derivative. I understand I'm the bad guy because I got angry first, but maybe you would understand if you put hours into something just for someone to click block without even testing or checking out the code, or a thought into what the changes they are suggesting mean. I'm not even sure I want to contribute this pr anymore this whole process has been so terrible since I left that comment on your pr.

fsnkty commented:
> > it's not about my time, the differences here seem unnecessary, it's not about wanting to avoid rebasing. blindly calling the advice of others to be a bad suggestion isn't productive. just looking to save us all some time n effort ^^```` but responses like this are more frustrating than productive, I'll leave you to it but please don't let this enable you to hold up the other PR later on whenever it's ready.. I'd still like to just use it in it's entirety rather than mushing this seemingly pointlessly diverged work together
> 
> What do you want to call a suggestion I strongly disagree with, that blocks my entire pr and makes my work a waste of time? I came up with this pr by myself as something I wanted to improve to the nix ecosystem. 

some stronger justification for why you disagree with it, it's not a personal thing.

> There's no diverging of work, just because you made the pr first doesn't mean my work is a derivative.

they aim for the same thing with different scopes, you diverge from existing work, no intentions to imply it's a derivative though I have suggested that maybe it should be in some ways.

> I understand I'm the bad guy because I got angry first, but maybe you would understand if you put hours into something just for someone to click block without even testing or checking out the code, or a thought into what the changes they are suggesting mean. 

these are some rude assumptions and don't invite collaboration :(

> I'm not even sure I want to contribute this pr anymore this whole process has been so terrible since I left that comment on your pr.
> 
> 

if your goal truely is to have the basics of the module included in nixpkgs faster, rather than focusing on the few blockers on the rest of it, that's fine! it's always nice to get stuff started and to get more work included.
it's just how you went about it and the reactions to others questions and suggestions that are frustrating, you haven't said anything on why my work isn't reusable here or how it wouldnt simplify the work for all of us ( you, me, reviewers ).
I'd like to hear your reasoning etc

Redhawk18 commented:
> if your goal truely is to have the basics of the module included in nixpkgs faster, rather than focusing on the few blockers on the rest of it, that's fine! it's always nice to get stuff started and to get more work included. it's just how you went about it and the reactions to others questions and suggestions that are frustrating, you haven't said anything on why my work isn't reusable here or how it wouldnt simplify the work for all of us ( you, me, reviewers ). I'd like to hear your reasoning etc

Because by the time I saw it a majority of it is already implemented, I'm not sure about the hardening still I would have to do a lot of testing to see if it even works well enough to include without breaking features in unexpected ways. Also are you sure you can still have an undeclarative setup if thats what the user desires or is that impossible with your pr? Can you blame me for thinking your pr was dead, there was over a month with no comments or new pushes?

fsnkty commented:
> > if your goal truely is to have the basics of the module included in nixpkgs faster, rather than focusing on the few blockers on the rest of it, that's fine! it's always nice to get stuff started and to get more work included. it's just how you went about it and the reactions to others questions and suggestions that are frustrating, you haven't said anything on why my work isn't reusable here or how it wouldnt simplify the work for all of us ( you, me, reviewers ). I'd like to hear your reasoning etc
> 
> Because by the time I saw it a majority of it is already implemented

not the kinda justification I'm looking for, that's very personal and a "but I did this work" nothing to do with why it is or isn't better than existing work.

> I'm not sure about the hardening still I would have to do a lot of testing to see if it even works well enough to include without breaking features in unexpected ways. 

testing is a requirement, you shouldn't want to include work you're unsure of, you also shouldn't want to include work you're unsure about the safety of.
I've done my own testing with the hardening specified in my PR, but yeah qbit is a big thing and I'm not 100% that I haven't missed something, this is what time and testing is for. please do share ya findings if anything of note

> Also are you sure you can still have an undeclarative setup if thats what the user desires or is that impossible with your pr?

not using the option is an option yeah, this discussion belongs on the other PR

>  Can you blame me for thinking your pr was dead, there was over a month with no comments or new pushes?
> 
> 

currently just do not have access to the hardware required to test or work on it 😕 not really justification to dismiss it for consideration in conversation when having your work reviewed 

remains unproductive, sorry I'll be dismissing this, best of luck!

Redhawk18 commented:
> remains unproductive, sorry I'll be dismissing this, best of luck!

My pr is better because I am active and ready to take suggestions and make changes to merge this. You are projecting your reasons you like your pr more onto me.

Redhawk18 commented:
@ambroisie Are you willing to change your mind about the name change before I do testing on the hardening? I rather ask now and close this than waste more of my time on a pr that won't get merged because it wasn't first.

Lassulus commented:
@Redhawk18 I think you need to calm down a little. The other PR shows still some activity and the author has shown a good deal of patience to communicate to you what are the expectations of the PR and what steps needs to be done to carry that over to the finish line. While I appreciate your motivation and energy to carry this over the finish line, just actively demanding peoples attention in an open source project is bad netiquette. We are a big project with a lot of contributors and sadly sometimes changes take time and cooperation to get into the right shape. NixOS gives you a lot of tools to locally test and develop changes without relying on upstream changes so immediate action is seldom necessary.

fsnkty commented:
> @fsnkty blocking me is rude after telling me my work on this pr has no value. I will be talking to the nixos owners team about this and your pr.

I blocked you after struggling too much to take the conversation somewhere productive, you're personal attacks to the expectations and assumptions of others and myself make me uncomfortable working with you.
you'll be staying blocked, threatening to take this to the "owners team" is just more hostility, continued no effort to resolve anything before involving others

mweinelt commented:
To clarify, @Lassulus response is on behalf of the so-called owners. We are the moderation team and we stand behind this message.

Redhawk18 commented:
> To clarify, @Lassulus response is on behalf of the so-called owners. We are the moderation team and we stand behind this message.

My mistake, I asked on the nixos discord who can I talk to and they told me the nixos org owners, not the mod team.

Redhawk18 commented:
> @Redhawk18 I think you need to calm down a little. The other PR shows still some activity and the author has shown a good deal of patience to communicate to you what are the expectations of the PR and what steps needs to be done to carry that over to the finish line. While I appreciate your motivation and energy to carry this over the finish line, just actively demanding peoples attention in an open source project is bad netiquette. We are a big project with a lot of contributors and sadly sometimes changes take time and cooperation to get into the right shape. NixOS gives you a lot of tools to locally test and develop changes without relying on upstream changes so immediate action is seldom necessary.

Thank you for responding, its my mistake for expecting things to move as fast as they do at my job, this is a totally different environment. That was just how I've been use to doing things. In my head the thought of creating a well maintained and well developed module for a service shouldn't belong in my config alone, that's why I wanted to have a pr.

I would like to extend an olive branch to @fsnkty

I believe both of us felt attacked by the other pr, I know I did. I reported you to the mod team because I felt this was a destructive road this was going down. I would appreciate it if you would accept that trying to merge my pr is in anyway hostile. Also I would like to add how my personal feelings tolds you do not matter in the context of pull requests, you can have coworkers you strongly dislike. However you still have to work as a team and work out or work around the differences you might have. 

I would like to close this pr if you unblock me and I can add some of the features I like from this pr to yours as suggestions. I believe this is a solution that would make everyone happy.

fsnkty commented:
> > @Redhawk18 I think you need to calm down a little. The other PR shows still some activity and the author has shown a good deal of patience to communicate to you what are the expectations of the PR and what steps needs to be done to carry that over to the finish line. While I appreciate your motivation and energy to carry this over the finish line, just actively demanding peoples attention in an open source project is bad netiquette. We are a big project with a lot of contributors and sadly sometimes changes take time and cooperation to get into the right shape. NixOS gives you a lot of tools to locally test and develop changes without relying on upstream changes so immediate action is seldom necessary.
> 
> Thank you for responding, its my mistake for expecting things to move as fast as they do at my job, this is a totally different environment. That was just how I've been use to doing things. In my head the thought of creating a well maintained and well developed module for a service shouldn't belong in my config alone, that's why I wanted to have a pr.
> 
> I would like to extend an olive branch to @fsnkty
> 
> I believe both of us felt attacked by the other pr, I know I did. I reported you to the mod team because I felt this was a destructive road this was going down. I would appreciate it if you would accept that trying to merge my pr is in anyway hostile. Also I would like to add how my personal feelings tolds you do not matter in the context of pull requests, you can have coworkers you strongly dislike. However you still have to work as a team and work out or work around the differences you might have.
> 
> I would like to close this pr if you unblock me and I can add some of the features I like from this pr to yours as suggestions. I believe this is a solution that would make everyone happy.
> 
> @Lassulus @mweinelt

thankfully in this environment you are not made to deal with all co contributors, I see no conflict to resolve as you are as free to work on this as you always were, I am in no way able or even willing to block you from doing so.
you are as welcome to contribute your understanding / potential findings / etc related to qbit or my pr more specifically as always.
blocking you was not a malicious or defensive thing, just didn't want notifications in my inbox from a discussion I couldn't see becoming productive.
my attempts to understand and make suggestions for your pr clearly have not worked, none of it was ever intended to be an attack to you or your work. this is why I didn't want notifications, I just assumed I was not helpful.

I cant manage to word my concerns properly so I'll unfortunately leave it at that for now and deal with any potential future issues whenever I'm actually able to lol.

best of luck, qbit is bizarre software.

Redhawk18 commented:
> I cant manage to word my concerns properly so I'll unfortunately leave it at that for now and deal with any potential future issues whenever I'm actually able to lol.
> 
> best of luck, qbit is bizarre software.

Communication skills are more important than any technical skill you could ever obtain, my pr is an example of that. My best wishes to your pr I hope you can make it before 30.05 releases.
